File: javascript/solutions/exactchange2.js
// Problem: https://open.kattis.com/problems/exactchange2

// Proper solution using dynamic programming. I originally did this in python3, but also did a JS port to compare execution.
// This is a variation of a well known set of problems, just Google "coin change problem", or visit some of the sites that helped me:
// http://stackoverflow.com/questions/3947867/find-the-least-number-of-coins-required-that-can-make-any-change-from-1-to-99-ce
// http://www.geeksforgeeks.org/find-minimum-number-of-coins-that-make-a-change/

let coins = [];
let test_cases_no = parseInt(readline(), 10);
while (test_cases_no > 0) {
  // table[x] will hold the minimum number of coins required for price x
  // Initialize all values in the table with the maximum value, which is the maximum price + 1
  
  // The fastest way I've found to init an array of lentgh 19999 with all values set to 10001
  // Execution time on kattis with this method - ~0.10 s
  let table = new Array(19999);
  for (let i = 0, value = 10001, size = 19999; i < size; i++) {
    table[i] = value;
  }
  
  // Another way, a bit slower, to init an array of lentgh 19999 with all values set to 10001
  // Execution time on kattis with this method - ~0.21 s
  // const table = [];
  // let size = 19999;
  // while(size--) table[size] = 10001;

  // if price is 0 ...
  table[0] = 0;

  // the price we need to pay
  const price = parseInt(readline());
  // the number of coins
  const coins_no = parseInt(readline());

  for (let i = 0; i < coins_no; i += 1) {
    coins.push(parseInt(readline()));
  }

  for (let i = 0; i < coins_no; i += 1) {
    let crt_price = price - 1;
    while (crt_price >= 0) {
      if (table[crt_price] === 10001) {
        crt_price -= 1;
        continue;
      }
      
      if (table[crt_price + coins[i]] > table[crt_price] + 1) {
        table[crt_price + coins[i]] = table[crt_price] + 1;
      }

      crt_price -= 1;
    }
  }

  let i = price
  while (true) {
    if (table[i] != 10001) {
      break;
    }
    i += 1;
  }

  print(i + ' ' + table[i]);
  
  test_cases_no -= 1;
  coins = [];
}

File: python3/solutions/exactchange2.py
# Problem: https://open.kattis.com/problems/exactchange2

########## GENERATE TEST CASES LOCALLY ##########
dummy_input = [
  '1',
  '1400',
  '3',
  '500',
  '1000',
  '2000',
]


def input():
  if len(dummy_input) != 0:
    return dummy_input.pop(0)
  raise EOFError()

################## CODE #####################

# (Untested) Naive solution using recursion to compute all possible combinations, see my other solution using dynamic programming.
# This is a variation of a well known set of problems, just Google "coin change problem", or visit some of the sites that helped me:
# http://stackoverflow.com/questions/3947867/find-the-least-number-of-coins-required-that-can-make-any-change-from-1-to-99-ce
# http://www.geeksforgeeks.org/find-minimum-number-of-coins-that-make-a-change/

def get_combinations(coins, length):
  # result = [] # without yield
  coins_len = len(coins)
  for i in range(coins_len):
    if length == 1:
      yield (coins[i],)
      # result.append([coins[i]]) # without yield
    else:
      for next in get_combinations(coins[i + 1:coins_len], length - 1):
        yield (coins[i],) + next
        # result.append([coins[i]] + next) # without yield

  # return result # without yield

coins = []
test_cases_no = int(input())
while test_cases_no > 0:
  price = int(input())
  coins_no = int(input())

  for i in range(coins_no):
    coins.append(int(input()))

  smallest_sum = 19999
  coins_used = 101
  for i in range(coins_no):
    if smallest_sum == price:
      break
    for combination in list(get_combinations(coins, i + 1)):
      combination_sum = sum(combination)

      if combination_sum == price:
        smallest_sum = combination_sum
        coins_used = i + 1
        break

      if combination_sum > price and combination_sum < smallest_sum:
        smallest_sum = combination_sum
        coins_used = i + 1

  print(smallest_sum, coins_used)

  test_cases_no -= 1
  coins = []
